### ISSUE TITLE
fit_generator using use_multiprocessing=True does not work on Windows 8.1 x64, python 3.5

### ISSUE THREAD

**Issue by erwinkendo** [keras-team/keras#10842]
Dear Keras community

I have been using keras succesfully for many tasks.

After implementing a custom data generator using the keras `Sequence `class, I tried using the `use_multiprocessing=True` of the `fit_generator` function, with more than 1 worker (so data can be fed to my GPU).

Unfortunately, after testing this setup in 3 different machines, the code seems to work only on Linux (even having a different GPU). 

- Behaviour using my windows 8.1 machine with python 3.5 installed using conda: Program freezes and do not start loading or training.
- Behaviour using my archlinux laptop, same software setup as the windows machine using conda: Program does work, feeds the data from disk, allows multiple threads and uses GPU for training.

Is this the expected behaviour on a windows machine?

Kind regards,

**Comment by bjoernholzhauer** [comment]
This [issue](https://github.com/matterport/Mask_RCNN/issues/13) is a bit unclear to me about whether that's expected or whether it should only occur with a standard data generator, but not with a sequence. I also get 

> ValueError: Using a generator with use_multiprocessing=True is not supported on Windows (no marshalling of generators across process boundaries). Instead, use single thread/process or multithreading.

with both (generator or sequence) when using multiprocessing=True in keras 2.2.0 in Python 3.6.6 under Windows 10 64-bit. I am not sure how to do multithreading in keras instead, either (the error message seems to suggest that might solve the problem, but I was unable to find any information exactly how one would do that).

**Comment by Dref360** [comment]
We do mimic Windows in the unit test using the 'spawn' method. So it "should" work.
I don't own a Windows machine so I can't really help you.

**Comment by bjoernholzhauer** [comment]
Can you point me to some code I should try, which would be expected to work? Does one need to do anything specific to use the 'spawn' method?

**Comment by Dref360** [comment]
Try running the tests in tests/keras/utils/data_utils.py

'spawn' (fork-exec) is the default on Windows. UNIX uses **fork** by default.
In those tests, we try with both mechanisms.

**Comment by bjoernholzhauer** [comment]
I will test it from home with my proper system. I just realized that with a sequence generator at least multi_processing=False and workers=4 (or so) does actually do multi-threading. I had missed that, because my toy example did not spend enough time in the sequence generator to make it really obvious. multi_processing=True of course still hangs the system.

**Comment by bjoernholzhauer** [comment]
@Dref360 Sorry, finally got around to trying it. What file do I need to provide the data_utils.py?

**Comment by bjoernholzhauer** [comment]
Any suggestions from anyone how to test this on a windows system (I'm honestly not clear on what files are needed for the test script @Dref360 pointed to)?

**Comment by loheden** [comment]
This still seems to be an issue. When use_multithreading=True, it is just hanging and literally nothing is happening. I am running it on Windows 10.

Setting workers to a number that is bigger than 1 seems to improve the speed even if use_multithereading=False. Why is this setup improving it?

Additionally, I also wonder if one needs to make its class generator (with `Sequence`) thread safe? Since I am not able to set use_multithreading to True, I am then wondering if I need to make my generator thread safe? In that case, I am also wondering if the thread safe version will give the desired performance improvement..?

**Comment by gitrdonator** [comment]
@Dref360 Couldn't the keras team update `keras/utils/data_utils.py` to pass a regular `multiprocessing.Lock()` at `Pool` creation time, using the `initializer` kwarg? This will make your lock instance global in all the child workers. See this Stackoverflow answer: [Python sharing a lock between processes](https://stackoverflow.com/questions/25557686/python-sharing-a-lock-between-processes).

**Comment by Dref360** [comment]
What would be the purpose of this Lock?

**Comment by gitrdonator** [comment]
@Dref360 To overcome the error on Windows that `TypeError: can't pickle _thread.lock objects`

**Comment by gitrdonator** [comment]
@Dref360 I create a generator class that extends `tensorflow.python.keras.preprocessing.image.Iterator`. I only implement `__init__` and `_get_batches_of_transformed_samples`. The problem is `Iterator` itself contains a `threading.Lock()`

```
def __init__(self, n, batch_size, shuffle, seed):
        ....
        self.lock = threading.Lock()
        ...
```

and uses it in its `next()` function to control index generation

```
def next(self):
        """For python 2.x.

        # Returns
            The next batch.
        """
        with self.lock:
            index_array = next(self.index_generator)
        # The transformation of images is not under thread lock
        # so it can be done in parallel
        return self._get_batches_of_transformed_samples(index_array)
```

When I try to use my generator and pass it to `fit_generator`, I inevitably get the error `TypeError: can't pickle _thread.lock objects`. I think locks can be marshalled in Linux, but not Windows (not sure).

My initial thought was to have `Sequence` hold a `self.lock`, then update `init_pool` and `init_pool_generator` in `data_utils.py` to accept a lock, changing the first two lines to be

```
    global _SHARED_SEQUENCES, _LOCK
    _SHARED_SEQUENCES = seqs
    _LOCK = lk
```

and lastly update the `_get_executor_init` functions in `SequenceEnqueuer` subclasses to add `lk` to `initargs`

```
return lambda seqs: mp.Pool(workers,
                                    initializer=init_pool,
                                    initargs=(seqs,lk))
```

Then, per the above mentioned StackOverflow answer, Iterator would inherit its lock from Sequence, but I think we would run into the same problem because `seqs` gets passed to initargs, which would then contain locks.

Could we create a separate class that just holds a threading lock and use multiple inheritance? i.e. Have `Iterator` extend `(Sequence, SomeLockContainerClass)`?

**Comment by gitrdonator** [comment]
@erwinkendo @bjoernholzhauer @Dref360 @loheden This cannot work in Windows. The original printout that 

```
ValueError: Using a generator with use_multiprocessing=True is not supported on Windows (no marshalling of generators across process boundaries). Instead, use single thread/process or multithreading.
```

must be added back in.

**Comment by gitrdonator** [comment]
@Dref360 Dumb question, but is there any way the thread locking could be moved to a function outside the `Iterator` class in and have `next` call it, similar to what was done with `init_pool` and `seqs` in `data_utils.py`? Or maybe just making turning `_flow_index` into a Queue of indeces to be accessed across processes?

**Comment by mchaniotakis** [comment]
Same here..! any solutions?

**Comment by Dref360** [comment]
PRs are welcome. I cannot work on this issue as I do not use Windows.

**Comment by txyugood** [comment]
I think the threading lock in the multiprocess module is unused. Can I remove the threading lock in the class Sequence?

**Comment by Dref360** [comment]
Sorry, but I don't see which Lock you're talking about.

In the class Sequence there is no lock: https://github.com/keras-team/keras/blob/master/keras/utils/data_utils.py#L305.


We do have a Lock hold by the _SEQUENCE_COUNTER:
https://github.com/keras-team/keras/blob/master/keras/utils/data_utils.py#L450


and internally there is a Lock inside the Queue. Both of these are really important.


COuld you point me to the code you're refering to?

________________________________
De : txyugood <notifications@github.com>
Envoyé : 6 août 2019 10:54:15
À : keras-team/keras <keras@noreply.github.com>
Cc : Frédéric Branchaud-Charron <Frederic.Branchaud-Charron@USherbrooke.ca>; Mention <mention@noreply.github.com>
Objet : Re: [keras-team/keras] fit_generator using use_multiprocessing=True does not work on Windows 8.1 x64, python 3.5 (#10842)


I think the threading lock in the multiprocess module is unused. Can I remove the threading lock in the class Sequence?

**Comment by txyugood** [comment]
I'm sorry, the threading lock in the class Iterator.
https://github.com/keras-team/keras-preprocessing/blob/master/keras_preprocessing/image/iterator.py#L43.

I mean the multiprocess module doesn't use the thread, it uses the process.
So is the threading lock necessary in the process ? 
Can I remove it, when I used the multiprocess on windows?

**Comment by Dref360** [comment]
I guess this is used when we do not use the OrderedEnqueuer, because we cannot iterate a generator by two deferent threads. If it really solves your problem, we can make the lock lazily in next.

________________________________
De : txyugood <notifications@github.com>
Envoyé : 6 août 2019 11:42:23
À : keras-team/keras <keras@noreply.github.com>
Cc : Frédéric Branchaud-Charron <Frederic.Branchaud-Charron@USherbrooke.ca>; Mention <mention@noreply.github.com>
Objet : Re: [keras-team/keras] fit_generator using use_multiprocessing=True does not work on Windows 8.1 x64, python 3.5 (#10842)


I'm sorry, the threading lock in the class Iterator.
https://github.com/keras-team/keras-preprocessing/blob/master/keras_preprocessing/image/iterator.py#L43.

I mean the multiprocess module doesn't use the thread, it uses the process.
So is the threading lock necessary in the process ?
Can I remove it, when I used the multiprocess on windows?

**Comment by txyugood** [comment]
Thank you for your reply. I investigated the code in 
https://github.com/keras-team/keras/blob/master/keras/engine/training_generator.py
https://github.com/keras-team/keras/blob/master/keras/utils/data_utils.py

You're right, the threading lock in the class Iterator is used by GeneratorEnqueuer(for the class that is not subclass of the Sequence).

But there is a bug in fit_generator\evaluate_generator\predict_generator method when we use the class ImageDataGenerator. The code ` is_sequence = isinstance(generator, Sequence)` is in `training_generator.py` of version 2.2.4. 
The class keras.utils.data_utils.Sequence  and tensorflow.python.keras.utils.data_utils.Sequence are detected as different class, so the is_sequence is set as False, so the GeneratorEnqueuer is used.This has been fixed in the master branch.

When we use the ImageDataGenerator(subclass of Sequence and the is_sequence  is True) and enable the multiprocess.The OrderedEnqueuer will be used, then the threading lock is not necessary, can we add some conditions to disable the threading lock?
I hava try this and  the problem `TypeError: can't pickle _thread.lock objects` was disappeard.

**Comment by evrial** [comment]
Still no solution?

**Comment by gitrdonator** [comment]
@evrial @txyugood @mchaniotakis I dont think there is a solution, unfortunately. Tensorflow has historically been built for Linux OSs. The `multiprocesssing` module is the best bet to get this working with Windows, but IMO it would take a complete rewrite of `preprocessing.image`.

**Comment by adam-grant-hendry** [comment]
I have a proposed "solution" that may interest others. Please note this is not a direct solution to the problem, but I believe a useful workaround. Please note this is coming from my experience with Tensorflow 1.15 (I have yet to use version 2). Please also see StackOverflow question [Is the class generator (inheriting Sequence) thread safe in Keras/Tensorflow?](https://stackoverflow.com/questions/52932406/is-the-class-generator-inheriting-sequence-thread-safe-in-keras-tensorflow/63641535#63641535)

# **TL;DR**

**_Install `wsl` version 2 on Windows, install Tensorflow in a Linux environment (e.g. Ubuntu) here, and then set `use_multiprocessing` to `True` to get this to work._**

**NOTE:** _The Windows Subshell for Linux (WSL) version 2 is only available in Windows 10, Version 1903, Build 18362 or higher._ **Be sure to upgrade your Windows version in Windows Update to get this to work.**

# **Long Answer**

For `multitasking` and `multithreading` (i.e. `parallelism` and `concurrency`), there are two operations we must consider:

 - `forking` = a parent process creates a copy of itself (a child) that has an exact copy of all the memory segments it uses
 - `spawning` = a parent process creates an entirely new child process that does not share its memory and the parent process must wait for the child process to finish before continuing

Linux supports `forking`, but Windows does not. Windows only supports `spawning`.

The reason Windows hangs when using `use_multiprocessing=True` is because the Python  `threading` module uses `spawn` for Windows. Hence, the parent process waits forever for the child to finish because the parent cannot transfer its memory to the child, so the child doesn't know what to do.

**On Windows, `use_multiprocessing=True` It is not `threadsafe`.** On Windows, if you've ever attempted to use a data generator or sequence, you've probably seen an error like this

```
ValueError: Using a generator with use_multiprocessing=True is not supported on Windows 
(no marshalling of generators across process boundaries). Instead, use single 
thread/process or multithreading.
```

`marshalling` means "transforming the memory representation of an object into a data format that is suitable for transmission." The error is saying that unlike Linux, which uses `fork`, `use_multiprocessing=True` doesn't work on Windows because it uses spawn` and cannot transfer its data to the child thread.

At this point, you may be asking yourself:

_"Wait...What about the Python Global Interpreter Lock (GIL)?..If Python only allows one thread to run at a time, why does it even have the `threading` module and why do we care about this in Tensorflow??!"_

The answer lies in the difference between `CPU-bound tasks` and `I/O-bound tasks`:

  - `CPU-bound tasks` = those that are waiting for data to be crunched
  - `I/O-bound tasks` = those that are waiting for input or output from other processes (i.e. data transferring)

In programming, when we say two tasks are `concurrent`, we mean they can start, run, and complete in overlapping time. When we say they are `parallel`, we mean they are literally running at the same time.

So, the GIL prevents threads from running _in parallel_, but not _concurrently_. The reason this is important for Tensorflow is because _concurrency_ is all about I/O operations (data transfer). A good dataflow pipeline in Tensorflow should try to be `concurrent` so that there's no lag time when data is being transferred to-and-from the CPU, GPU, and/or RAM and training finishes faster. (Rather than have a thread sit and wait until it gets data back from somewhere else, we can have it executing image preprocessing or something else until the data gets back.)

------------------------------------------------------------------------------------------
#### **IMPORTANT ASIDE:** The `GIL` was made in Python because everything in Python is an object. (This is why you can do "weird" things with "dunder/magic" methods, like `(5).__add__(3)` to get `8`

#### **NOTE:** In the above, parentheses are needed around `5` since `5.` is a `float`, so we need to take advantage of order of operations by using parentheses.

#### Python handles memory and garbage collection by counting all references made to individual objects. When the count goes to 0, Python deletes the object. If two threads tried to access the same object simultaneously, or if one thread finishes faster than another, you can get a `race condition` and objects would be deleted "randomly". We could put a `lock` on each thread, but then we would be unable to prevent `deadlocks`.

#### Losing `parallel` thread execution was seen by Guido (and by myself, though it is certainly arguable) as a minor loss because we still maintained I/O concurrent operations, and tasks could still be run in `parallel` by running them on different cpu cores (i.e. `multiprocessing`). Hence, this is (one reason) why Python has both the `threading` and `multiprocessing` modules.
------------------------------------------------------------------------------------------

Now back to `threadsafe`. When running `concurrent`/`parallel` tasks, you have to watch out for additional things. Two big ones are:

1. `race conditions` - operations don't take exactly the same time to compute each time a program as run (why with `timeit` we average over a number of runs). Because threads will finish at different times depending on the run, you get different results with each run.

1. `deadlock` - if two threads try to access the same memory at the same time, you'll get an error. To prevent this, we add a `lock` or `mutex` (mutual exclusion) to threads to prevent other threads from accessing the same memory while it is running. However, if two threads need to access the same memory, are locked, and each thread depends on the other one finishing in order to execute, the program hangs.

I bring this up because Tensorflow needs to be able to `pickle` Python objects to make code run faster. (`pickling` is turning objects and data into byte code, much in the same way that an entire program's source code is converted into an `exe` on Windows). The Tensorflow `Iterator.__init__()` method locks threads and contains a `threading.Lock()`

```
def __init__(self, n, batch_size, shuffle, seed):
    ...
    self.lock = threading.Lock()
    ...
```

The problem is Python cannot `pickle` threading lock objects on Windows (i.e. Windows cannot `marshall` thread locks to `child` threads).

If you were to try to use a generator and pass it to `fit_generator`, you will get the error

```
TypeError: can't pickle _thread.lock objects
```

So, while `use_multiprocessing=True` is threadsafe on Linux, it is not on Windows.

**Solution:** Around June 2020, Microsoft came out with version 2 of the Windows Subshell for Linux (`wsl`). This was significant because it enabled GPU hardware acceleration. Version 1 was "simply" a driver between Windows NT and Linux, whereas `wsl` is now actually a kernel. Thus, you can now install Linux on Windows, open a bash shell from the command prompt, and (most importantly) access hardware. Thus, it is now possible to install `tensorflow-gpu` on `wsl`. In addition, you'll now be able to use `fork`.

**Thus, I recommend 

1. Installing `wsl` version 2 on Windows and add your desired Linux environment
1. Install `tensorflow-gpu` in a virtual environment in `wsl` Linux environment here
1. Retry `use_multiprocessing=True` to see if it works.**

_CAVEAT:_ I haven't tested this yet to verify that it works, but to the best of my limited knowledge, I believe it should.


  [1]: https://github.com/keras-team/keras/issues/10842
  [2]: https://www.youtube.com/watch?v=SxOsJPaxHME

**Comment by josuerocha** [comment]
I am still able to reproduce the issue on Python 3.9.0 and Tensorflow 2.6.0.

 I tried WSL 2 but the speedup relative to Windows without multiprocessing was of just 20%. 

Is there any alternative or solution ?

**Comment by rayjennings3rd** [comment]
I am trying to get this same option to work on MacOS Monterey. While other python packages are able to use multi-processing, I see no improvement at all with this option. I am running 8 cores, tensorflow 2.6. I am using Keras Sequential models. I have:

```
workers=6,
use_multiprocessing=True
```

And there is zero difference between having this on or off.

**Comment by iyadahmed** [comment]
I had the same problem, to fix it you just need to wrap your code with `if __name__ == "__main__"`
https://docs.python.org/3/library/multiprocessing.html#multiprocessing-programming:~:text=Safe%20importing%20of%20main%20module
![image](https://user-images.githubusercontent.com/25552173/172709711-53497759-22ec-4f77-ad82-4d3211550436.png)
